src/reflect/scala/reflect/internal/Types.scala

```diff
@@ -359,7 +359,12 @@ trait Types
     /** Map a constant type or not-null-type to its underlying base type,
      *  identity for all other types.
      */
-    def deconst: Type = this
+    def deconst: Type = {
+      val deconst0 = new TypeMap {
+        def apply(tp: Type): Type = tp.deconst
+      }
+      mapOver(deconst0)
+    }
```

Comment:
Could the deconsting logic be turned into a `TypeMap` rather than a virtual call? This allocates at least one object per `deconst` call. Even having a single `DeconstMap` with this logic could arguably be nicer. (A downside being that you'd need to click through to see what it's doing.)

Reply:
Yeah, this looks a bit wrong, and looks like it has exponential complexity.

I think the intent is:

```
object deconstDeep extends TypeMap {
   def apply(tp: Type) = tp.deconst.mapOver(this)
}
```

That would leave the shallow `deconst` in place and introduce a new operation. You'd need to decide at each call to `deconst` which was appropriate.

But I don't think this is the right approach, and would prefer to prevent the constant type from getting into the args of a type ref in the first place.

This is the smallest change to fix provided test case:

https://github.com/scala/scala/compare/2.13.x...retronym:topic/deconst-type-arg

We should review `AsSeenFromMap` to see if there are other cases to test (type members rather than params, maybe?), and find the suitable place for the shallow deconst.

Reply:
Ouch ... yes, the exponential bit is a tad embarrassing. I was trying to preserve the existing internal method while reusing `TypeMaps` traversal logic. I'll move to an external `TypeMap` and replace `.deconst` calls case by case.

@retronym your fix doesn't go quite far enough. Although it correctly handles cases of the first form (including the reify/unreify case), examples of the second form `test/files/pos/t10768.scala` go back to being a crashers. Also note that constant types can be smuggled in in other ways than via a `TypeRef`, eg. via a refinement, and that these things can be arbitrarily nested. I think that deep deconsting _is_ necessary, at least at some of the sites where currently only shallow deconsting is done.

Reply:
@retronym here's a case which escapes your fix,
```scala
object Test {
  abstract class Box {
    type T
    val t: T
    def foo: T = {
      println("effect")
      t
    }
  }

  object Box {
    def apply(x: String): Box { type T = x.type } = new Box {
      type T = x.type
      val t = x
    }
  }

  def main(args: Array[String]): Unit = {
    val bar = Box("foo").foo
  }
}
```

Reply:
@retronym I missed your comment about `AsSeenFromMap` earlier. It strikes me that the deconst logic is pretty much orthogonal to the as seen from logic ... we're not relativising types based on point of view, we're making a decision about constant folding and inlining. I appreciate that the kind of traversal that `AsSeenFromMap` is doing might pick up the problematic inlining cases, but if it does that seems more or less accidental.

I also noticed that there's no handling of refinements `AsSeenFromMap` ... I'm guessing that that's deliberate. The new test I added shows that we do need to recurse through refinements to deconst properly though. I don't see how to (easily) reconcile that difference.

Reply:
I'm going to push a bit harder on the shallow approach. Here's my next attempt, that handles `run/t10788*.scala`: https://github.com/scala/scala/compare/2.13.x...retronym:topic/deconst-type-arg?expand=1

Reply:
Just tweaked that to handle poly types and nested method types.